README.rst
===========
Python XTEA
===========

.. image:: https://api.codeclimate.com/v1/badges/563d6ad2607d6ed6fda3/maintainability
   :target: https://codeclimate.com/github/tgates42/xtea/maintainability
   :alt: Maintainability

.. image:: https://api.codeclimate.com/v1/badges/563d6ad2607d6ed6fda3/test_coverage
   :target: https://codeclimate.com/github/tgates42/xtea/test_coverage
   :alt: Test Coverage

.. image:: https://travis-ci.org/tgates42/xtea.svg?branch=master
   :target: https://travis-ci.org/tgates42/xtea
   :alt: Travis CI build status

This is an XTEA-Cipher implementation in Python (eXtended Tiny Encryption Algorithm).

    XTEA is a blockcipher with 8 bytes blocksize and 16 bytes Keysize (128-Bit).
    The algorithm is secure at 2014 with the recommend 64 rounds (32 cycles). This
    implementation supports following modes of operation:
    ECB, CBC, CFB, OFB, CTR


Example:

    >>> from xtea4 import *
    >>> key = " "*16  # Never use this
    >>> text = "This is a text. "*8
    >>> x = new(key, mode=MODE_OFB, IV="12345678")
    >>> c = x.encrypt(text)
    >>> text == x.decrypt(c)
    True
    
Note
====
   
    I do NOT guarantee that this implementation (or the base cipher) is secure. If you find bugs, please report them at
    https://github.com/tgates42/xtea/issues . 



changelog.rst
Changelog
---------

(dev) Version 0.6.1; ...
~~~~~~~~~~~~~~~~~~~~~~~~

 - Improved tests
 - PEP8-style formatting
 - Unittests: Counter, modes (but not results of them!), test vectors
 - [BREAKING CHANGE] Counter class is now in xtea4.counter

(unreleased) Version 0.6.0; Oct 16, 2016
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Python 3 does work now
 - [BREAKING CHANGE] counters cannot return numbers any more, they must return bytestrings now
 - [BREAKING CHANGE] Cipher objects remember state, so two consecutive calls to XTEACipher.encrypt should not return the same
 - improved documentation

(unreleased) Version 0.5.0; Oct 15, 2016
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Removed CBCMAC

Version 0.4.1; Jul 30, 2015
~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Fixed installer

Version 0.4.0; Jul 12, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Buggless & PEP compliant CTR
 - CTR mode works with strings now
 - raises DeprecatedWarning if a number is returned
 - CBCMAC class added (use static method CBCMAC.new(args) to create)

Version 0.3.2; Jul 11, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Minor Fixes

Version 0.3.1; Jul 11, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~~

 -  Minor Fixes
 - Fixed that the length of data will not be checked

Version 0.3.0; Jul 11, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Added CFB mode
 - Fully working with PEP 272
 - Raising NotImplementedError only on PGP-CFB (OpenPGP) mode
 - Wheel support and changelog (0.2.1)

Version 0.2.1 - dev; Jul 10, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Never released...

 - Added better wheel support for uploading (just for me) with a setup.cfg
 - Added this file (auto uploading on pypi/warehouse and github)
 - (upload.py for github)

Version 0.2.0; Jul 9, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~

 - Added a test feature; warning in CTR

 - Added a test feature
 - Raises warning on CTR, added a handler that CTR will not crash anymore ;) 

Version 0.1.1; Jul 9, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~~

[0.1.1] NotImplementedError on CFB

 - Module raises a NotImplementedError on CFB
 - Minor changes

Version 0.1; Jun 22, 2014
~~~~~~~~~~~~~~~~~~~~~~~~~

[0.1] Initial release

 - Supports all mode except CFB
 - Buggy CTR ( "ß" = "\\xc3\\x9f" )
 - Working with PEP 272, default mode is ECB


tests/test_xtea.py
import unittest
import os

from time import clock

from xtea4 import MODE_CBC, MODE_CTR, MODE_ECB, MODE_OFB, MODE_CFB, XTEACipher
from xtea4.counter import Counter

AMOUNT = 50


def _test_mode(mode):
    plain = os.urandom(56) * 8
    counter = Counter(os.urandom(8))
    key = os.urandom(16)
    iv = os.urandom(8)

    e = XTEACipher(key, IV=iv, counter=counter, mode=mode)
    encrypted = e.encrypt(plain)

    d = XTEACipher(key, IV=iv, counter=counter, mode=mode)

    counter.reset()
    decrypted = d.decrypt(encrypted)
    if plain != decrypted:
        raise AssertionError("Invalid decryption!")


class TestModes(unittest.TestCase):
    def testECB(self):
        print("Testing ECB")

        start = clock()
        for i in range(AMOUNT):
            _test_mode(MODE_ECB)
        end = clock()
        time = end - start
        print("Time: %s" % str(round(time, 3)))

    def testCBC(self):
        print("Testing CBC")
        start = clock()

        for i in range(AMOUNT):
            _test_mode(MODE_CBC)

        end = clock()
        time = end - start
        print("Time: %s" % str(round(time, 3)))

    def testCFB(self):
        print("Testing CFB")
        start = clock()

        for i in range(AMOUNT):
            _test_mode(MODE_CFB)

        end = clock()
        time = end - start
        print("Time: %s" % str(round(time, 3)))

    def testOFB(self):
        print("Testing OFB")
        start = clock()

        for i in range(AMOUNT):
            _test_mode(MODE_OFB)

        end = clock()
        time = end - start
        print("Time: %s" % str(round(time, 3)))

    def testCTR(self):
        print("Testing CTR")
        start = clock()

        for i in range(AMOUNT):
            _test_mode(MODE_CTR)

        end = clock()
        time = end - start
        print("Time: %s" % str(round(time, 3)))


if __name__ == "__main__":
    unittest.main()


tests/test_counter.py
import unittest

from xtea4.counter import Counter, to_bytes, from_bytes


class TestIntToBytes(unittest.TestCase):
    def testLittleEndian(self):
        result = to_bytes(0x04030201, 4, 'little')
        expected = b'\x01\x02\x03\x04'
        self.assertEqual(result, expected)

    def testBigEndian(self):
        result = to_bytes(0x01020304, 4, 'big')
        expected = b'\x01\x02\x03\x04'
        self.assertEqual(result, expected)

    def testStrangeByteOrder(self):
        with self.assertRaises(ValueError):
            to_bytes(100, 4, 'banana')


class TestBytesToInt(unittest.TestCase):
    def testLittleEndian(self):
        result = from_bytes(b'\x01\x02\x03\x04', 'little')
        expected = 0x04030201
        self.assertEqual(result, expected)

    def testBigEndian(self):
        result = from_bytes(b'\x01\x02\x03\x04', 'big')
        expected = 0x01020304
        self.assertEqual(result, expected)

    def testStrangeByteOrder(self):
        with self.assertRaises(ValueError):
            from_bytes(b'\x01\x02\x03\x04', 'banana')


class TestCounter(unittest.TestCase):
    def testLittleEndian(self):
        counter = Counter(b'\x00\x00\x00\x00\x00\x00\x00\x00', 'little')
        self.assertEqual(counter(), b'\x00\x00\x00\x00\x00\x00\x00\x00')
        self.assertEqual(counter(), b'\x01\x00\x00\x00\x00\x00\x00\x00')

    def testBigEndian(self):
        counter = Counter(b'\x00\x00\x00\x00\x00\x00\x00\x00', 'big')
        self.assertEqual(counter(), b'\x00\x00\x00\x00\x00\x00\x00\x00')
        self.assertEqual(counter(), b'\x00\x00\x00\x00\x00\x00\x00\x01')

    def testMaxValue(self):
        counter = Counter(b'\xff\xff\xff\xff\xff\xff\xff\xff', 'little')
        self.assertEqual(counter(), b'\xff\xff\xff\xff\xff\xff\xff\xff')
        self.assertEqual(counter(), b'\x00\x00\x00\x00\x00\x00\x00\x00')


if __name__ == "__main__":
    unittest.main()


tests/test_vectors.py
#!/usr/bin/env python3
"""
Testvectors for XTEA found at:
 - http://www.tayloredge.com/reference/Mathematics/XTEA.pdf
 
"""

import xtea
from binascii import hexlify
from binascii import unhexlify


def _unwrap(s):
    return unhexlify(s.replace(' ', ''))


TEST_VECTORS = [
    {
        'k': '27f917b1 c1da8993 60e2acaa a6eb923d',
        'p': 'af20a390 547571aa',
        'c': 'd26428af 0a202283'
    },
    {
        'k': '31415926 53589793 23846264 33832795',
        'p': '02884197 16939937',  # 48 digits of PI
        'c': '46e2007d 58bbc2ea',
    },
    {
        'k': '1234abc1 234abc12 34abc123 4abc1234',
        'p': 'abc1234a bc1234ab',
        'c': '5c0754c1 f6f0bd9b',
    },
    {
        'k': 'abc1234a bc1234ab c1234abc 1234abc1',
        'p': '234abc12 34abc123',
        'c': 'cdfcc72c 24bc116b',
    },
    {
        'k': 'deadbeef deadbeef deadbeef deadbeef',
        'p': 'deadbeef deadbeef',
        'c': 'faf28cb5 0940c0e0',
    },
    {
        'k': 'deadbeef deadbeef deadbeef deadbeef',
        'p': '9647a918 9ec565d5',
        'c': 'deadbeef deadbeef',
    },
]


def test_encryption():
    for i in range(len(TEST_VECTORS)):
        vector = TEST_VECTORS[i]
        e = _unwrap(vector['p'])
        c = _unwrap(vector['c'])
        k = _unwrap(vector['k'])

        x = xtea4.new(key=k, mode=xtea4.MODE_ECB)
        try:
            r = x.encrypt(e)
            assert r == c
        except AssertionError:
            print("Error on test %s:" % i)
            print("Expected %s, got %s!" % (hexlify(c).decode(),
                                            hexlify(r).decode()))
            raise


def test_decryption():
    for i in range(len(TEST_VECTORS)):
        vector = TEST_VECTORS[i]
        e = _unwrap(vector['p'])
        c = _unwrap(vector['c'])
        k = _unwrap(vector['k'])

        x = xtea4.new(key=k, mode=xtea4.MODE_ECB)
        try:
            r = x.decrypt(c)
            assert r == e
        except AssertionError:
            print("Error on test %s:" % i)
            print("Expected %s, got %s!" % (hexlify(e).decode(),
                                            hexlify(r).decode()))
            raise


if __name__ == "__main__":
    if not __debug__:
        raise Exception("This script uses \"assert\" for tests. "
                        "Do not -O or -OO to run it.")
    test_encryption()
    test_decryption()


xtea4/counter.py
import struct
import sys

PY_3 = sys.version_info.major >= 3

if PY_3:

    def to_bytes(integer, length, byteorder):
        return integer.to_bytes(length, byteorder)

    def from_bytes(bytesarray, byteorder):
        return int.from_bytes(bytesarray, byteorder)
else:

    def to_bytes(integer, length, byteorder='big'):
        h = '%x' % integer
        s = ('0' * (len(h) % 2) + h).zfill(length * 2).decode('hex')
        if byteorder == 'big':
            return s
        elif byteorder == 'little':
            return s[::-1]
        else:
            raise ValueError("byteorder must be either 'little' or 'big'")

    def from_bytes(bytesarray, byteorder):
        if len(bytesarray) == 4:
            size = 'L'
        elif len(bytesarray) == 8:
            size = 'Q'

        if byteorder == 'big':
            return struct.unpack(">" + size, bytesarray)[0]
        elif byteorder == 'little':
            return struct.unpack("<" + size, bytesarray)[0]
        else:
            raise ValueError("byteorder must be either 'little' or 'big'")


class Counter:
    """Small counter for CTR mode, based on arrays
    Example:
    
        >>> from xtea4.counter import Counter
        >>> nonce = b"$2dUI84e" # This should be random
        >>> c = Counter(nonce)
        >>> c()
        b'%2dUI84e'
        >>> c()
        b'&2dUI84e'
        >>> c()
        b"'2dUI84e"
        >>> c.reset()
        >>> c()
        b'%2dUI84e'
    """

    def __init__(self, nonce, byteorder='big'):
        """Constructor for a counter which is suitable for CTR mode.

        nonce (bytes): The start value;
                it MUST be unique to be secure.
                The secrets module or os.urandom(n) are solid
                choices for generating random bytes.
        """
        self.__nonce = nonce
        self.byteorder = byteorder
        self.__current = from_bytes(self.__nonce, self.byteorder)

    def __call__(self):
        """Increase the counter by one.

        Returns:
            bytes
        """
        value = to_bytes(self.__current, 8, self.byteorder)
        self.__current += 1
        self.__current %= 2**64
        return value

    def reset(self):
        """Reset the counter to the nonce."""

        self.__current = from_bytes(self.__nonce, self.byteorder)


xtea4/__init__.py
"""
XTEA-Cipher in Python (eXtended Tiny Encryption Algorithm)

XTEA is a blockcipher with 8 bytes blocksize and 16 bytes Keysize (128-Bit).
The algorithm is secure at 2014 with the recommend 64 rounds (32 cycles). This
implementation supports following modes of operation:
ECB, CBC, CFB, OFB, CTR

Example:

>>> from xtea4 import *
>>> key = " "*16  # Never use this
>>> text = "This is a text. "*8
>>> x = new(key, mode=MODE_OFB, IV="12345678")
>>> c = x.encrypt(text)
>>> c.encode("hex")
'fa66ec11b82e38bc77c14be093bb8aa0d7fe0fb9e6ec015
7a22d254fee43aea9a64c8dbb2c2b899f66800f264419c8e
8796ad8f94c7758b916428019d10573943324a9dcf60f883
1f0f925cd7215e5dd4f1334d9ee242d41ac02d0a64c49663
e5897bfd2450982379267e6cd7405b477ccc19d6c0d32e2f
887b76fe01d621a8d'
>>> text == x.decrypt(c)
True
"""

from __future__ import print_function

import struct
import binascii
import sys
import warnings

from .counter import Counter

MODE_ECB = 1
MODE_CBC = 2
MODE_CFB = 3
MODE_PGP = 4
MODE_OFB = 5
MODE_CTR = 6

PY_3 = sys.version_info.major >= 3

if PY_3:

    def b_ord(n):
        return n

    def b_chr(n):
        return bytes([n])
else:

    def b_ord(n):
        return ord(n)

    def b_chr(n):
        return chr(n)


block_size = 64
key_size = 128


def new(key, **kwargs):
    """Create an "XTEACipher" object.
    It fully PEP-272 comliant, default mode is ECB.
    Args:
        key (bytes): The key for encrytion/decryption. Must be 16 in length

    Kwargs:

        mode (int): Mode of operation, must be one of this::
            1 = ECB
            2 = CBC
            3 = CFB
            5 = OFB
            6 = CTR
        
        IV (bytes): Initialisation vector (needed with CBC/CFB).
            Must be 8 in length.
        
        counter (callable object): a callable counter wich returns bytestrings

            .. versionchanged:: 0.5.0
               Only callable objects returning bytestrings can be used,
               previously integers instead of callables were allowed, too.
        
        endian (char / string):
            how data is beeing extracted (default "!" = big endian)
            ..seealso:: modules :py:mod:`struct`

        rounds (int / float): How many rounds are going to be used,
            one round are two cycles, there are no *half* cycles.
            The minimum secure rounds are 37 (default 64)

    Raises:
        ValueError if invalid/not all data is given.
        NotImplementedError on MODE_PGP

    Returns:
       XTEACipher object

    """
    return XTEACipher(key, **kwargs)


################ XTEACipher class


class XTEACipher(object):
    """The cipher class

    Functions:
    encrypt -- encrypt data
    decrypt -- decrypt data

    _block -- splits the data in blocks (you may need padding)

    Constants:
    block_size = 8

    Variables:
    IV -- the initialisation vector (default None or "\00"*8)
    counter -- counter for CTR (default None)
    
    """

    block_size = 8
    IV = None
    counter = None

    def __init__(self, key, **kwargs):
        """\
        Alternative constructor.
        Create an cipher object.

        Args:
            key (bytes): The key for encrytion/decryption. Must be 16 in length

        Kwargs:
            mode (int): Mode of operation, must be one of this::

                1 = ECB
                2 = CBC
                3 = CFB
                5 = OFB
                6 = CTR

            IV (bytes): Initialisation vector (needed with CBC/CFB).
                Must be 8 in length.

            counter (callable object): a callable counter wich returns bytes
                or int (needed with CTR)
      
            endian (char / string):
                how data is beeing extracted (default "!")
                ..seealso:: modules :py:mod:`struct`

        Raises:
            ValueError if invalid/not all required data is give.
            NotImplementedError on MODE_PGP.

        Creates:
            XTEACipher object

        """

        self.key = key
        if len(key) != key_size / 8:  # Check key len
            raise ValueError("Key must be 128 bit long")

        keys = kwargs.keys()  # arguments

        self.mode = kwargs.get("mode")

        if self.mode == None:
            self.mode = MODE_ECB  # if not given
            warnings.warn("Using implicit ECB!")

        if self.mode not in (
                MODE_ECB, MODE_CBC, MODE_CFB, MODE_OFB, MODE_CTR):
            raise NotImplementedError(
                "The selected mode of operation does not exist.")

        self.IV = kwargs.get("IV")

        # cbc, cfb and ofb need an iv
        if self.mode in (MODE_CBC, MODE_CFB, MODE_OFB) and (
                self.IV is None or len(self.IV) != self.block_size):
            raise ValueError(
                "CBC, CFB and OFB need an IV with a length of 8 bytes!")

        self.counter = kwargs.get("counter")

        if self.mode == MODE_CTR and not callable(self.counter): 
            raise ValueError("CTR mode needs a callable counter")

        self.rounds = int(kwargs.get("rounds", 64))
        self.endian = kwargs.get("endian", "!")

        if self.mode == MODE_OFB:

            def keygen():
                while True:
                    self.IV = _encrypt(self.key, self.IV, self.rounds // 2)
                    for k in self.IV:
                        yield b_ord(k)

            self._keygen = keygen()

        elif self.mode == MODE_CTR:

            def keygen():
                while True:
                    self.IV = _encrypt(self.key,
                                       self.counter(), self.rounds // 2)
                    for k in self.IV:
                        yield b_ord(k)

            self._keygen = keygen()

    def encrypt(self, data):
        """\
        Encrypt data, it must be a multiple of 8 in length except for
        CTR and OFB mode of operation. When using the OFB or CTR mode, the
        function for encryption and decryption is the same.

        Args:
            data (bytes): The data to encrypt.
        Returns:
            bytestrings
        Raises:
            ValueError
        """

        if self.mode in (MODE_OFB, MODE_CTR):
            return self._stream(data)

        args = (self.rounds // 2, self.endian)

        out = []
        blocks = self._block(data)

        for block in blocks:
            if self.mode == MODE_ECB:
                ecd = _encrypt(self.key, block, *args)
            elif self.mode == MODE_CBC:
                xored = xor_strings(self.IV, block)
                ecd = self.IV = _encrypt(self.key, xored, *args)
            elif self.mode == MODE_CFB:
                keystream = _encrypt(self.key, self.IV, *args)
                ecd = self.IV = xor_strings(keystream, block)
            else:
                raise ValueError("Unknown mode of operation!")

            out.append(ecd)

        return b"".join(out)


    def decrypt(self, data):
        """\
        Decrypt data, it must be a multiple of 8 in length except for
        CTR and OFB mode of operation. When using the OFB or CTR mode, the
        function for encryption and decryption is the same.

        Args:
            data (bytes): The data to decrypt.
        Returns:
            bytestrings
        Raises:
            ValueError
        """

        if self.mode in (MODE_OFB, MODE_CTR):
            return self._stream(data)

        args = (self.rounds // 2, self.endian)

        out = []
        blocks = self._block(data)

        for block in blocks:
            if self.mode == MODE_ECB:
                dec = _decrypt(self.key, block, *args)
            elif self.mode == MODE_CBC:
                decrypted_but_not_xored = _decrypt(self.key, block, *args)
                dec = xor_strings(self.IV, decrypted_but_not_xored)
                self.IV = block
            elif self.mode == MODE_CFB:
                keystream = _encrypt(self.key, self.IV, *args)
                dec = xor_strings(keystream, block)
                self.IV = block
            else:
                raise ValueError("Unknown mode of operation!")

            out.append(dec)

        return b"".join(out)


    def _stream(self, data):
        xor = [b_chr(x ^ y) for (x, y) in zip(map(b_ord, data), self._keygen)]
        return b"".join(xor)

    def _block(self, s):
        l = []
        rest_size = len(s) % self.block_size
        if rest_size:
            raise ValueError("Input string must be a multiple of blocksize "
                             "in length")
        for i in range(len(s) // self.block_size):
            l.append(s[i * self.block_size:((i + 1) * self.block_size)])
        return l


################ Util functions: basic encrypt/decrypt, OFB, xor, stringToLong
"""
This are utilities only, use them only if you know what you do.

Functions:
_encrypt -- Encrypt one single block of data.
_decrypt -- Decrypt one single block of data.
xor_strings -- xor to strings together.
"""


def _encrypt(key, block, n=32, endian="!"):
    """Encrypt one single block of data.

    Only use if you know what to do.

    Keyword arguments:
    key -- the key for encrypting (and decrypting)
    block  -- one block plaintext
    n -- cycles, one cycle is two rounds, more cycles
          -> more security and slowness (default 32)
    endian -- how struct will handle data (default "!" (big endian/network))
    """
    v0, v1 = struct.unpack(endian + "2L", block)
    k = struct.unpack(endian + "4L", key)
    sum, delta, mask = 0, 0x9e3779b9, 0xffffffff
    for round in range(n):
        v0 = (v0 + (((v1 << 4 ^ v1 >> 5) + v1) ^ (sum + k[sum & 3]))) & mask
        sum = (sum + delta) & mask
        v1 = (v1 +
              (((v0 << 4 ^ v0 >> 5) + v0) ^ (sum + k[sum >> 11 & 3]))) & mask
    return struct.pack(endian + "2L", v0, v1)


def _decrypt(key, block, n=32, endian="!"):
    """Decrypt one single block of data.

    Only use if you know what to do.

    Keyword arguments:
    key -- the key for encrypting (and decrypting)
    block  -- one block ciphertext
    n -- cycles, one cycle is two rounds, more cycles =
          -> more security and slowness (default 32)
    endian -- how struct will handle data (default "!" (big endian/network))
    """
    v0, v1 = struct.unpack(endian + "2L", block)
    k = struct.unpack(endian + "4L", key)
    delta, mask = 0x9e3779b9, 0xffffffff
    sum = (delta * n) & mask
    for round in range(n):
        v1 = (v1 -
              (((v0 << 4 ^ v0 >> 5) + v0) ^ (sum + k[sum >> 11 & 3]))) & mask
        sum = (sum - delta) & mask
        v0 = (v0 - (((v1 << 4 ^ v1 >> 5) + v1) ^ (sum + k[sum & 3]))) & mask
    return struct.pack(endian + "2L", v0, v1)


if PY_3:

    def xor_strings(s, t):
        """xor to strings together.

        Keyword arguments:
        s -- string one
        t -- string two
        """
        return bytes([(x ^ y) for x, y in zip(s, t)])
else:

    def xor_strings(s, t):
        """xor to strings together.

        Keyword arguments:
        s -- string one
        t -- string two
        """
        return "".join(chr(ord(x) ^ ord(y)) for x, y in zip(s, t))
